kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/ConfigWatcherService.java

```diff
@@ -0,0 +1,330 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.kroxylicious.proxy.internal;
+
+import java.io.IOException;
+import java.nio.file.FileSystems;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardWatchEventKinds;
+import java.nio.file.WatchEvent;
+import java.nio.file.WatchKey;
+import java.nio.file.WatchService;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Consumer;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.kroxylicious.proxy.config.ConfigParser;
+import io.kroxylicious.proxy.config.Configuration;
+
+/**
+ * Service that monitors a configuration file for changes and triggers hot-reload when detected.
+ * Uses Java NIO WatchService for efficient file system monitoring.
+ */
+public class ConfigWatcherService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ConfigWatcherService.class);
+
+    private final Path configFilePath;
+    private final ConfigParser configParser;
+    private final Consumer<Configuration> onConfigurationChanged;
+    private final Duration debounceDelay;
+
+    private final AtomicBoolean running = new AtomicBoolean(false);
+    private final ScheduledExecutorService watcherExecutor = Executors.newSingleThreadScheduledExecutor(r -> {
+        Thread t = new Thread(r, "config-watcher");
+        t.setDaemon(true);
+        return t;
+    });
+    private final ScheduledExecutorService reloadExecutor = Executors.newSingleThreadScheduledExecutor(r -> {
+        Thread t = new Thread(r, "config-reload");
+        t.setDaemon(true);
+        return t;
+    });
+
+    private WatchService watchService;
+    private WatchKey watchKey;
+    private Configuration currentConfiguration;
+    private Instant lastModified = Instant.EPOCH;
```

Comment:
Since you're using executors in this class, let's at least add comments about which threads are supposed to be reading and writing these non-final fields. It's the sort of thing which is correct on day one, but someone later on gets a bit careless because the things that made is correct are a bit subtle.